src/hydrosheds.cc

```diff
@@ -6,96 +6,306 @@
+            return result;
+        }
+        DownstreamIterationHelper followstream(const hydrosheds::HydroshedsDataSet& dataset, hydrosheds::Coordinate x)
+        {
+            DownstreamIterationHelper result{dataset, x};
+            return result;
+        }
+/*
+        // DownstreamIteratorHelper
+        DownstreamIterator DownstreamIterationHelper::begin()
+        {
+            auto feature = dataset.layer->GetFeature(0);
+            RiverSegment segment(dataset.layer, feature);
+            hydrosheds::Coordinate closest = segment.getStartingPoint();
+            // iterate over all from the FullDatasetRiverSegmentIterator and find RiverSegment closest to x
+            for(auto i : dataset)
+            {
+                auto start = i.getStartingPoint();
+                if(impl::norm(x, start)<impl::norm(x, closest))
+                {
+                    segment = i;
+                    closest = i.getStartingPoint();
+                }
+            }
+            DownstreamIterator result(segment);
+            return result;
+        }
+*/
+        // ----------------------------
+
+
+        // DownstreamIteratorHelper end();
+        DownstreamIterator DownstreamIterationHelper::end()
+        {
+            /*
+            auto segment = dataset.getSegment();
+            for(int i = 0; 1; i++)
+            {
+                if (!segment.hasDownstreamSegment())
+                {
+                    DownstreamIterator end(segment);
+                    return end;
+                }
+            }
+             */
+        }
+
+
+        // ----------------------------
+
+        // DowmstreamIterator
+        DownstreamIterator DownstreamIterator::operator++()
+        {
+            auto next = this->segment.getDownstreamSegment();
+            this->segment = next;
+            return (*this);
+        }
+
+        DownstreamIterator DownstreamIterator::operator++(int)
+        {
+            auto result = (*this);
+            auto next = this->segment.getDownstreamSegment();
+            this->segment = next;
+            return result;
+        }
+        bool DownstreamIterator::operator==(const DownstreamIterator& a)
+        {
+            auto a_x = a.segment.getStartingPoint();
+            auto a_y = a.segment.getEndPoint();
+            auto _x = this->segment.getStartingPoint();
+            auto _y = this->segment.getEndPoint();
+            if(a_x[0]==_x[0] && a_x[1]==_x[1] && a_y[0]==_y[0] && a_y[1]==_y[1])
+            {
+                return true;
+            }else{
+                return false;
+            }
+        }
+        bool DownstreamIterator::operator!=(const DownstreamIterator& a)
+        {
+            auto a_x = a.segment.getStartingPoint();
+            auto a_y = a.segment.getEndPoint();
+            auto _x = this->segment.getStartingPoint();
+            auto _y = this->segment.getEndPoint();
+            if(a_x[0]==_x[0] && a_x[1]==_x[1] && a_y[0]==_y[0] && a_y[1]==_y[1])
```

Comment:
Equality comparison with floating point numbers are always dangerous, as they might not hold bitwise. Also, I think we have better ways of implementing this operation.